vm/systemSmartContracts/register.go

```diff
@@ -0,0 +1,52 @@
+package systemSmartContracts
+
+import (
+	"errors"
+	"github.com/ElrondNetwork/elrond-go/vm"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+	"math/big"
+)
+
+type registerSC struct {
+	eei        vm.SystemEI
+	stakeValue *big.Int
+}
+
+// NewRegisterSmartContract creates a register smart contract
+func NewRegisterSmartContract(stakeValue *big.Int, eei vm.SystemEI) (*registerSC, error) {
+	if stakeValue == nil {
+		return nil, errors.New("initial stake value is nil")
+	}
+	if eei == nil || eei.IsInterfaceNil() {
+		return nil, errors.New("environment interface is nil")
+	}
+
+	reg := &registerSC{stakeValue: big.NewInt(0).Set(stakeValue), eei: eei}
+	return reg, nil
+}
+
+// Execute calls one of the functions from the register smart contract and runs the code according to the input
+func (r *registerSC) Execute(args *vm.ExecuteArguments) vmcommon.ReturnCode {
+	if CheckIfNil(args) != nil {
+		return vmcommon.UserError
+	}
+
+	if args.Value.Cmp(r.stakeValue) != 0 {
+		return vmcommon.UserError
+	}
+
```

Comment:
I would have expected that the registration SC would have done the checking of the required balance of the sender. should have transferred toward the registration address the stake value and modify some internal values (that will then be sent toward VMOutput for permanent storage) that will track down the registration event. This vars might have been a map containing tuples like {address, metablock nonce, stake_value} or a variant like this.

Reply:
Balance is checked by smartContract processor and the minus part is taken out of the sender address balance before executing a smart contract. Here the registration has to write into a buffer/adapter/cache which will update after correct execution the peer state.